content/6_Advanced/Wavelet.mdx

```diff
@@ -32,9 +30,248 @@ Like a segment tree on values rather than indices.
 	</Resource>
 </Resources>
 
-### Solution - Range K-th Smallest
+Suppose you want to support the following queries:
+
+- Given a range $l$, $r$ count number of occurrences of value $x$.
+- Given a range $l$, $r$ find the $k$ smallest element
+
+With wavelet tree you can easily support those queries in $log(M)$ time.
+when M is the maximum value in the array.
+
+## Wavelet Tree Structure
+
+A wavelet tree is a binary tree where each node represents a range of values.
+The root node covers the entire range, and each subsequent level splits the
+range into two halves.
+
+## Defining the structure
+
+We are going to maintain a segment tree over values instead of indices.
+Each segment contains indices whose values lie within the segment's range.
+Let's save those indicies in a vector.
+Notice that a index can be in at most $log(M)$ segments
+
+<Spoiler title="Wavelet tree Visualization">
+
+Lets say our array is: $[3,5,3,1,2,2,3,4,5,5]$
+
+![Wavelet Tree Visualization](./assets/diagram.png)
+</Spoiler>
+
+## Solving the First Type of query
+**Given a range l, r count number of occurrences of value x.**
+
+To calculate the number of occurrences from $𝑙$ to $𝑟$, we can use the following
+formula:
+
+$$
+\begin{aligned}
+\texttt{occurrences}(l, r) = \texttt{occurrences}(r) - \texttt{occurrences}(l)
+\end{aligned}
+$$
+
+This reduces the problem to counting the number of occurrences in a prefix.
+
+### Approaching the Solution
+
+One way to solve the problem is to go to the leaf node
+and perform a binary search for the number of indices less than $𝑟$
+However, let's explore a different approach that can also be extended to the
+second type of query.
+
+### A different way
```

Comment:
you don't need to split up the module into this many headers

Reply:
Would using ### instead of ## also do the job? or would you prefer to erase some of the headers?

Reply:
i think it's best if you erase some of the headers and just split up your ideas into paragraphs

Reply:
unaddressed?

Reply:
I have erased some of the headers. Would you also like me to erase the first type of query and the second type of query header as well?